File: Sources/HoverPocket/PocketApps/PocketAppExecutionRuntime.swift
import Foundation

struct PocketAppWorkflowDraft: Equatable, Sendable {
    let packageID: String
    let workflowID: String
    let plan: CapabilityExecutionPlan
    let preparation: CapabilityBrokerPreparation
}

@MainActor
final class PocketAppExecutionRuntime {
    private static let presentableWorkflowCapabilities: Set<PocketCapabilityKey> = [
        PocketCapabilityKeys.timerStart,
        PocketCapabilityKeys.stickyUpsert
    ]

    let package: PocketAppPackage
    let userStateStore: PocketAppUserStateStore?

    private let broker: CapabilityBroker
    private let principal: CapabilityPrincipal
    private let grantedPermissions: Set<String>
    private let timeZone: TimeZone

    init(
        package: PocketAppPackage,
        broker: CapabilityBroker,
        userID: String,
        grantedPermissions: Set<String>,
        timeZone: TimeZone = .current,
        userStateStore: PocketAppUserStateStore? = nil
    ) {
        self.package = package
        self.broker = broker
        self.principal = CapabilityPrincipal(userID: userID, pocketAppID: package.manifest.id)
        self.grantedPermissions = grantedPermissions
        self.timeZone = timeZone
        self.userStateStore = userStateStore
    }

    func query(
        reference: String,
        arguments: [String: PocketJSONValue],
        now: Date = Date()
    ) async throws -> CapabilityObject {
        let key = try capabilityKey(reference)
        let request = try requestedCapability(key)
        guard request.effect == .pure || request.effect == .privateRead else {
            throw CapabilityBrokerError.invalidPlan("query_effect")
        }
        let resolvedArguments = try arguments.mapValues { try resolve($0, inputs: [:], now: now) }
        try validateScope(resolvedArguments, request: request)
        let nonce = UUID().uuidString.lowercased()
        let plan = CapabilityExecutionPlan(
            id: "pocket-query:\(nonce)",
            createdAt: now,
            origin: .pocketSurface,
            principal: principal,
            appContext: appContext,
            steps: [
                CapabilityPlanStep(
                    id: "query",
                    capability: key,
                    arguments: resolvedArguments,
                    idempotencyKey: "pocket-query.\(nonce)",
                    dependencies: []
                )
            ],
            requiredPermissions: request.permissions
        )
        let permissions = permissionSet
        let preparation = try broker.prepare(plan, permissions: permissions, now: now)
        guard preparation.approvalRequest == nil else {
            throw CapabilityBrokerError.invalidPlan("query_approval")
        }
        let receipt = try await broker.execute(
            plan,
            permissions: permissions,
            approvalGrant: nil,
            now: now
        )
        guard receipt.status == .succeeded,
              let output = receipt.steps.first?.output else {
            throw CapabilityBrokerError.unavailable(key)
        }
        return output
    }

    func prepare(
        workflowID: String,
        inputs: [String: CapabilityValue],
        now: Date = Date()
    ) throws -> PocketAppWorkflowDraft {
        guard let workflow = package.workflows[workflowID] else {
            throw CapabilityBrokerError.invalidPlan("pocket_workflow")
        }
        try validateInputs(inputs, workflow: workflow)
        let nonce = UUID().uuidString.lowercased()
        let steps = try workflow.steps.map { step in
            guard Self.supportsWorkflowPresentation(step.capability) else {
                throw CapabilityBrokerError.invalidPlan("pocket_workflow_presentation")
            }
            let resolvedArguments = try step.arguments.mapValues { try resolve($0, inputs: inputs, now: now) }
            let arguments = try Self.canonicalWorkflowArguments(
                resolvedArguments,
                capability: step.capability
            )
            let request = try requestedCapability(step.capability)
            try validateScope(arguments, request: request)
            return CapabilityPlanStep(
                id: step.id,
                capability: step.capability,
                arguments: arguments,
                idempotencyKey: "pocket-workflow.\(nonce).\(step.id)",
                dependencies: step.dependencies
            )
        }
        let plan = CapabilityExecutionPlan(
            id: "pocket-workflow:\(nonce)",
            createdAt: now,
            origin: .pocketSurface,
            principal: principal,
            appContext: appContext,
            steps: steps,
            requiredPermissions: workflow.requiredPermissions
        )
        return PocketAppWorkflowDraft(
            packageID: package.manifest.id,
            workflowID: workflowID,
            plan: plan,
            preparation: try broker.prepare(plan, permissions: permissionSet, now: now)
        )
    }

    func approveAndExecute(
        _ draft: PocketAppWorkflowDraft,
        now: Date = Date()
    ) async throws -> CapabilityWorkflowReceipt {
        try validateDraft(draft)
        guard let request = draft.preparation.approvalRequest else {
            throw CapabilityBrokerError.approvalRequired
        }
        let grant = try broker.decideApproval(
            requestID: request.id,
            planDigest: draft.preparation.planDigest,
            decision: .approve,
            now: now
        )
        return try await broker.execute(
            draft.plan,
            permissions: permissionSet,
            approvalGrant: grant,
            now: now
        )
    }

    func reject(_ draft: PocketAppWorkflowDraft, now: Date = Date()) {
        do {
            try validateDraft(draft)
            guard let request = draft.preparation.approvalRequest else { return }
            _ = try broker.decideApproval(
                requestID: request.id,
                planDigest: draft.preparation.planDigest,
                decision: .reject,
                now: now
            )
        } catch CapabilityBrokerError.approvalRejected {
        } catch {
        }
    }

    private var appContext: CapabilityAppContext {
        CapabilityAppContext(
            id: package.manifest.id,
            version: package.manifest.version,
            manifestDigest: package.manifestDigest
        )
    }

    private var permissionSet: CapabilityPermissionSet {
        let requestedPermissions = package.manifest.requestedCapabilities.reduce(into: Set<String>()) {
            $0.formUnion($1.permissions)
        }
        return CapabilityPermissionSet(
            principal: principal,
            permissions: grantedPermissions.intersection(requestedPermissions)
        )
    }

    private func requestedCapability(_ key: PocketCapabilityKey) throws -> PocketAppRequestedCapability {
        guard let request = package.manifest.requestedCapabilities.first(where: { $0.key == key }) else {
            throw CapabilityBrokerError.unknownCapability(key)
        }
        return request
    }

    private func validateDraft(_ draft: PocketAppWorkflowDraft) throws {
        guard draft.packageID == package.manifest.id,
              package.workflows[draft.workflowID] != nil,
              draft.plan.principal == principal,
              draft.plan.appContext == appContext,
              draft.plan.origin == .pocketSurface else {
            throw CapabilityBrokerError.invalidPlan("pocket_draft")
        }
    }

    private func validateInputs(
        _ inputs: [String: CapabilityValue],
        workflow: PocketAppWorkflowDocument
    ) throws {
        guard Set(inputs.keys) == Set(workflow.inputs.keys) else {
            throw CapabilityBrokerError.invalidPlan("pocket_inputs")
        }
        for (name, type) in workflow.inputs {
            guard let value = inputs[name], Self.accepts(value, type: type) else {
                throw CapabilityBrokerError.invalidPlan("pocket_input_\(name)")
            }
        }
    }

    private static func accepts(_ value: CapabilityValue, type: String) -> Bool {
        switch (type, value) {
        case ("string", .string), ("entity-ref", .string), ("integer", .integer),
             ("number", .integer), ("number", .number), ("boolean", .bool):
            true
        case ("date-time", .string(let value)):
            CapabilityDateCodec.date(from: value) != nil
        default:
            false
        }
    }

    private func resolve(
        _ value: PocketJSONValue,
        inputs: [String: CapabilityValue],
        now: Date
    ) throws -> CapabilityValue {
        switch value {
        case .null:
            return .null
        case .bool(let value):
            return .bool(value)
        case .number(let value):
            if value.rounded() == value, value >= Double(Int.min), value <= Double(Int.max) {
                return .integer(Int(value))
            }
            return .number(value)
        case .string(let value):
            if value.hasPrefix("$input.") {
                let name = String(value.dropFirst("$input.".count))
                guard let resolved = inputs[name] else {
                    throw CapabilityBrokerError.invalidPlan("pocket_binding")
                }
                return resolved
            }
            if value == "$context.timezone" {
                return .string(timeZone.identifier)
            }
            if value == "$context.todayFocusStableKey" {
                return .string("today-focus:\(Self.localDateKey(now, timeZone: timeZone))")
            }
            if value.hasPrefix("$") {
                throw CapabilityBrokerError.invalidPlan("pocket_context")
            }
            return .string(value)
        case .array(let values):
            return .array(try values.map { try resolve($0, inputs: inputs, now: now) })
        case .object(let object):
            return .object(try object.mapValues { try resolve($0, inputs: inputs, now: now) })
        }
    }

    private func validateScope(
        _ arguments: CapabilityObject,
        request: PocketAppRequestedCapability
    ) throws {
        guard case .object(let scope)? = request.scope else { return }
        if case .string(let range)? = scope["range"], arguments["range"] != .string(range) {
            throw CapabilityBrokerError.invalidPlan("pocket_scope_range")
        }
        if case .string(let namespace)? = scope["namespace"] {
            guard case .string(let stableKey)? = arguments["stableKey"],
                  stableKey.hasPrefix("\(namespace):") else {
                throw CapabilityBrokerError.invalidPlan("pocket_scope_namespace")
            }
        }
    }

    private func capabilityKey(_ reference: String) throws -> PocketCapabilityKey {
        guard let marker = reference.lastIndex(of: "@"),
              let version = Int(reference[reference.index(after: marker)...]),
              version >= 1 else {
            throw CapabilityBrokerError.invalidPlan("pocket_capability")
        }
        return PocketCapabilityKey(id: String(reference[..<marker]), version: version)
    }

    private static func localDateKey(_ date: Date, timeZone: TimeZone) -> String {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = timeZone
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }

    private static func canonicalWorkflowArguments(
        _ arguments: CapabilityObject,
        capability: PocketCapabilityKey
    ) throws -> CapabilityObject {
        var canonical = arguments
        if capability == PocketCapabilityKeys.timerStart {
            guard case .string(let title)? = canonical["title"] else {
                throw CapabilityBrokerError.invalidPlan("pocket_workflow_presentation")
            }
            canonical["title"] = .string(TodayFocusApprovalText.sanitize(title))
        } else if capability == PocketCapabilityKeys.stickyUpsert {
            guard case .string(let title)? = canonical["title"],
                  case .string(let body)? = canonical["body"] else {
                throw CapabilityBrokerError.invalidPlan("pocket_workflow_presentation")
            }
            canonical["title"] = .string(TodayFocusApprovalText.sanitize(title))
            canonical["body"] = .string(TodayFocusApprovalText.sanitize(body))
        }
        return canonical
    }

    static func supportsWorkflowPresentation(_ capability: PocketCapabilityKey) -> Bool {
        presentableWorkflowCapabilities.contains(capability)
    }
}

File: windows/src/HoverPocket.Shell/PocketApps/PocketAppExecutionRuntime.cs
using System.Globalization;
using System.Text.Json;
using HoverPocket.Shell.Capabilities;

namespace HoverPocket.Shell.PocketApps;

internal sealed record PocketAppWorkflowDraft(
    string PackageId,
    string WorkflowId,
    CapabilityExecutionPlan Plan,
    CapabilityBrokerPreparation Preparation);

internal sealed class PocketAppExecutionRuntime : IDisposable
{
    private static readonly HashSet<PocketCapabilityKey> PresentableWorkflowCapabilities =
    [
        CapabilityIds.TimerStart,
        CapabilityIds.StickyUpsert
    ];

    private readonly CapabilityBroker _broker;
    private readonly CapabilityPrincipal _principal;
    private readonly IReadOnlySet<string> _permissions;
    private readonly TimeZoneInfo _timeZone;
    private readonly PocketAppActivationLease? _activationLease;
    private bool _disposed;

    public PocketAppExecutionRuntime(
        PocketAppPackage package,
        CapabilityBroker broker,
        string userId,
        IReadOnlySet<string> grantedPermissions,
        TimeZoneInfo? timeZone = null,
        PocketAppUserStateStore? userStateStore = null,
        PocketAppActivationLease? activationLease = null)
    {
        Package = package;
        _broker = broker;
        _principal = new CapabilityPrincipal(userId, package.Manifest.Id);
        _timeZone = timeZone ?? TimeZoneInfo.Local;
        var requested = package.Manifest.RequestedCapabilities
            .SelectMany(item => item.Permissions)
            .ToHashSet(StringComparer.Ordinal);
        _permissions = grantedPermissions.Where(requested.Contains).ToHashSet(StringComparer.Ordinal);
        UserStateStore = userStateStore;
        _activationLease = activationLease;
    }

    public PocketAppPackage Package { get; }

    public PocketAppUserStateStore? UserStateStore { get; }

    internal bool IsActivationActive => _activationLease?.IsActive ?? true;

    internal void EnsureActivationActive() => _activationLease?.RequireActive();

    public void Dispose()
    {
        if (_disposed) { return; }
        _disposed = true;
        UserStateStore?.Dispose();
    }

    public async Task<JsonElement> QueryAsync(
        string reference,
        JsonElement arguments,
        DateTimeOffset? now = null,
        CancellationToken cancellationToken = default)
    {
        EnsureActivationActive();
        var key = CapabilityKey(reference);
        var request = RequestedCapability(key);
        if (request.Effect is not (CapabilityEffect.Pure or CapabilityEffect.PrivateRead))
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "query_effect");
        }
        var current = now ?? DateTimeOffset.Now;
        var resolved = ResolveObject(arguments, new Dictionary<string, JsonElement>(StringComparer.Ordinal), current);
        ValidateScope(resolved, request);
        var nonce = Guid.NewGuid().ToString("D");
        var plan = new CapabilityExecutionPlan(
            $"pocket-query:{nonce}",
            current,
            CapabilityOrigin.PocketSurface,
            _principal,
            AppContext,
            [new CapabilityPlanStep("query", key, resolved, $"pocket-query.{nonce}", [])],
            request.Permissions);
        var permissions = PermissionSet;
        var preparation = _broker.Prepare(plan, permissions, current);
        if (preparation.ApprovalRequest is not null)
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "query_approval");
        }
        using var linkedCancellation = LinkedActivationCancellation(cancellationToken);
        var effectiveCancellation = linkedCancellation?.Token ?? cancellationToken;
        effectiveCancellation.ThrowIfCancellationRequested();
        var receipt = await _broker.ExecuteAsync(plan, permissions, null, current, effectiveCancellation);
        EnsureActivationActive();
        return receipt.Status == CapabilityReceiptStatus.Succeeded
            && receipt.Steps.FirstOrDefault()?.Output is { } output
            ? output
            : throw new CapabilityBrokerException("CAPABILITY_UNAVAILABLE", key.Id);
    }

    public PocketAppWorkflowDraft Prepare(
        string workflowId,
        IReadOnlyDictionary<string, JsonElement> inputs,
        DateTimeOffset? now = null)
    {
        EnsureActivationActive();
        if (!Package.Workflows.TryGetValue(workflowId, out var workflow))
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_workflow");
        }
        ValidateInputs(inputs, workflow);
        var current = now ?? DateTimeOffset.Now;
        var nonce = Guid.NewGuid().ToString("D");
        var steps = workflow.Steps.Select(step =>
        {
            if (!SupportsWorkflowPresentation(step.Capability))
            {
                throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_workflow_presentation");
            }
            var resolvedArguments = ResolveObject(
                CapabilityJson.From(step.Arguments.ToDictionary(item => item.Key, item => item.Value)),
                inputs,
                current);
            var arguments = CanonicalWorkflowArguments(resolvedArguments, step.Capability);
            var request = RequestedCapability(step.Capability);
            ValidateScope(arguments, request);
            return new CapabilityPlanStep(
                step.Id,
                step.Capability,
                arguments,
                $"pocket-workflow.{nonce}.{step.Id}",
                step.Dependencies);
        }).ToArray();
        var plan = new CapabilityExecutionPlan(
            $"pocket-workflow:{nonce}",
            current,
            CapabilityOrigin.PocketSurface,
            _principal,
            AppContext,
            steps,
            workflow.RequiredPermissions);
        return new PocketAppWorkflowDraft(
            Package.Manifest.Id,
            workflowId,
            plan,
            _broker.Prepare(plan, PermissionSet, current));
    }

    public async Task<CapabilityWorkflowReceipt> ApproveAndExecuteAsync(
        PocketAppWorkflowDraft draft,
        DateTimeOffset? now = null,
        CancellationToken cancellationToken = default)
    {
        EnsureActivationActive();
        ValidateDraft(draft);
        var request = draft.Preparation.ApprovalRequest
            ?? throw new CapabilityBrokerException("CAPABILITY_APPROVAL_REQUIRED", draft.Plan.Id);
        var current = now ?? DateTimeOffset.Now;
        var grant = _broker.DecideApproval(
            request.Id,
            draft.Preparation.PlanDigest,
            CapabilityApprovalDecision.Approve,
            current);
        using var linkedCancellation = LinkedActivationCancellation(cancellationToken);
        var effectiveCancellation = linkedCancellation?.Token ?? cancellationToken;
        effectiveCancellation.ThrowIfCancellationRequested();
        var receipt = await _broker.ExecuteAsync(draft.Plan, PermissionSet, grant, current, effectiveCancellation);
        EnsureActivationActive();
        return receipt;
    }

    public void Reject(PocketAppWorkflowDraft draft, DateTimeOffset? now = null)
    {
        if (!IsActivationActive) { return; }
        ValidateDraft(draft);
        if (draft.Preparation.ApprovalRequest is not { } request)
        {
            return;
        }
        try
        {
            _ = _broker.DecideApproval(
                request.Id,
                draft.Preparation.PlanDigest,
                CapabilityApprovalDecision.Reject,
                now ?? DateTimeOffset.Now);
        }
        catch (CapabilityBrokerException ex) when (ex.Code == "CAPABILITY_APPROVAL_REJECTED")
        {
        }
    }

    private CapabilityAppContext AppContext => new(
        Package.Manifest.Id,
        Package.Manifest.Version,
        Package.ManifestDigest);

    private CapabilityPermissionSet PermissionSet => new(_principal, _permissions);

    private CancellationTokenSource? LinkedActivationCancellation(CancellationToken cancellationToken) =>
        _activationLease is null
            ? null
            : CancellationTokenSource.CreateLinkedTokenSource(
                cancellationToken,
                _activationLease.CancellationToken);

    private PocketAppRequestedCapability RequestedCapability(PocketCapabilityKey key) =>
        Package.Manifest.RequestedCapabilities.FirstOrDefault(item => item.Key == key)
        ?? throw new CapabilityBrokerException("CAPABILITY_UNKNOWN", key.Id);

    private void ValidateDraft(PocketAppWorkflowDraft draft)
    {
        if (draft.PackageId != Package.Manifest.Id
            || !Package.Workflows.ContainsKey(draft.WorkflowId)
            || draft.Plan.Principal != _principal
            || draft.Plan.AppContext != AppContext
            || draft.Plan.Origin != CapabilityOrigin.PocketSurface)
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_draft");
        }
    }

    private static void ValidateInputs(
        IReadOnlyDictionary<string, JsonElement> inputs,
        PocketAppWorkflowDocument workflow)
    {
        if (!inputs.Keys.ToHashSet(StringComparer.Ordinal).SetEquals(workflow.Inputs.Keys))
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_inputs");
        }
        foreach (var input in workflow.Inputs)
        {
            if (!inputs.TryGetValue(input.Key, out var value) || !Accepts(value, input.Value))
            {
                throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", $"pocket_input_{input.Key}");
            }
        }
    }

    private static bool Accepts(JsonElement value, string type) => type switch
    {
        "string" or "entity-ref" => value.ValueKind == JsonValueKind.String,
        "integer" => value.ValueKind == JsonValueKind.Number && value.TryGetInt32(out _),
        "number" => value.ValueKind == JsonValueKind.Number,
        "boolean" => value.ValueKind is JsonValueKind.True or JsonValueKind.False,
        "date-time" => value.ValueKind == JsonValueKind.String
            && DateTimeOffset.TryParse(value.GetString(), CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out _),
        _ => false
    };

    private JsonElement ResolveObject(
        JsonElement value,
        IReadOnlyDictionary<string, JsonElement> inputs,
        DateTimeOffset now)
    {
        if (value.ValueKind != JsonValueKind.Object)
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_arguments");
        }
        return CapabilityJson.From(value.EnumerateObject().ToDictionary(
            property => property.Name,
            property => Resolve(property.Value, inputs, now),
            StringComparer.Ordinal));
    }

    private object? Resolve(
        JsonElement value,
        IReadOnlyDictionary<string, JsonElement> inputs,
        DateTimeOffset now)
    {
        return value.ValueKind switch
        {
            JsonValueKind.Null => null,
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.Number when value.TryGetInt32(out var integer) => integer,
            JsonValueKind.Number => value.GetDouble(),
            JsonValueKind.String => ResolveString(value.GetString() ?? string.Empty, inputs, now),
            JsonValueKind.Array => value.EnumerateArray().Select(item => Resolve(item, inputs, now)).ToArray(),
            JsonValueKind.Object => value.EnumerateObject().ToDictionary(
                property => property.Name,
                property => Resolve(property.Value, inputs, now),
                StringComparer.Ordinal),
            _ => throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_json")
        };
    }

    private object? ResolveString(
        string value,
        IReadOnlyDictionary<string, JsonElement> inputs,
        DateTimeOffset now)
    {
        if (value.StartsWith("$input.", StringComparison.Ordinal))
        {
            var name = value["$input.".Length..];
            return inputs.TryGetValue(name, out var input)
                ? Resolve(input, inputs, now)
                : throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_binding");
        }
        if (value == "$context.timezone")
        {
            return _timeZone.Id;
        }
        if (value == "$context.todayFocusStableKey")
        {
            return $"today-focus:{TimeZoneInfo.ConvertTime(now, _timeZone):yyyy-MM-dd}";
        }
        if (value.StartsWith('$'))
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_context");
        }
        return value;
    }

    private static void ValidateScope(JsonElement arguments, PocketAppRequestedCapability request)
    {
        if (request.Scope is not { ValueKind: JsonValueKind.Object } scope)
        {
            return;
        }
        if (scope.TryGetProperty("range", out var range)
            && (!arguments.TryGetProperty("range", out var actualRange)
                || actualRange.GetString() != range.GetString()))
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_scope_range");
        }
        if (scope.TryGetProperty("namespace", out var namespaceElement))
        {
            var expected = namespaceElement.GetString() ?? string.Empty;
            if (!arguments.TryGetProperty("stableKey", out var stableKey)
                || stableKey.ValueKind != JsonValueKind.String
                || PocketStableKey.Namespace(stableKey.GetString() ?? string.Empty) != expected)
            {
                throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_scope_namespace");
            }
        }
    }

    private static PocketCapabilityKey CapabilityKey(string reference)
    {
        var marker = reference.LastIndexOf('@');
        if (marker <= 0
            || !int.TryParse(reference[(marker + 1)..], NumberStyles.None, CultureInfo.InvariantCulture, out var version)
            || version < 1)
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_capability");
        }
        return new PocketCapabilityKey(reference[..marker], version);
    }

    private static JsonElement CanonicalWorkflowArguments(
        JsonElement arguments,
        PocketCapabilityKey capability)
    {
        var values = arguments.EnumerateObject().ToDictionary(
            property => property.Name,
            property => property.Value.Clone(),
            StringComparer.Ordinal);
        if (capability == CapabilityIds.TimerStart)
        {
            CanonicalizeVisibleString(values, "title");
        }
        else if (capability == CapabilityIds.StickyUpsert)
        {
            CanonicalizeVisibleString(values, "title");
            CanonicalizeVisibleString(values, "body");
        }
        else
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_workflow_presentation");
        }
        return CapabilityJson.From(values);
    }

    private static void CanonicalizeVisibleString(
        IDictionary<string, JsonElement> values,
        string field)
    {
        if (!values.TryGetValue(field, out var value) || value.ValueKind != JsonValueKind.String)
        {
            throw new CapabilityBrokerException("CAPABILITY_PLAN_INVALID", "pocket_workflow_presentation");
        }
        values[field] = CapabilityJson.From(TodayFocusApprovalText.Sanitize(value.GetString() ?? string.Empty));
    }

    internal static bool SupportsWorkflowPresentation(PocketCapabilityKey capability) =>
        PresentableWorkflowCapabilities.Contains(capability);
}
